runtime/include/cloe/data_broker.hpp

```diff
@@ -0,0 +1,1306 @@
+/*
+ * Copyright 2023 Robert Bosch GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+/**
+ * \file cloe/data_broker.cpp
+ * \see  cloe/data_broker.hpp
+ *
+ * Classes:
+ *   DataBroker:
+ *     A central registry for (type erased) signals.
+ *   
+ *   Signal:
+ *     A type-erased abstraction of a signal.
+ *  
+ *   Container<T>
+ *     An optional container storing the value of a signal and abstracting the interaction with Signal.
+ *   
+ * Background:
+ *   Real world simulations utilize manifolds of signals (variables) and coresponding types. 
+ *   Still the very basic problem around a single signal (variable) boils down to CRUD
+ *   - declaring a signal (Create)
+ *   - reading its value / receiving value-changed events (Read)
+ *   - writing its value / triggering value-changed events (Update)
+ *   - NOT NEEDED: undeclaring a signal (Delete)
+ *
+ * Concept:
+ *   Abstract CRU(D) operations by registering signals in a uniform way by their name.
+ */
+
+#pragma once
+#ifndef CLOE_DATA_BROKER_HPP_
+#define CLOE_DATA_BROKER_HPP_
+
+#include <any>
+#include <cassert>
+#include <functional>
+#include <map>
+#include <memory>
+#include <stdexcept>
+#include <vector>
+
+#include <fmt/format.h>
+
+namespace cloe {
+
+namespace databroker {
+
+template <typename TValue>
+using on_value_change_event_std_function_t = std::function<void(const TValue*)>;
+template <typename TValue>
+using on_value_change_event_function_t = void (*)(const TValue*);
+template <typename TClass, typename TValue>
+using on_value_change_event_member_function_t = void (TClass::*)(const TValue*);
+
+}  // namespace databroker
+
+/**
+ * Predicate which determines whether a type is wanted/unwanted as signal datatype
+ */
+template <typename TValue>
+constexpr bool is_wanted_type =
+    std::is_same<TValue, typename std::remove_cv<typename std::remove_pointer<
+                             typename std::remove_reference<TValue>::type>::type>::type>::value;
+/**
+ * SFINAE template to determine return-type for enabling functions consuming wanted type T
+ */
+template <typename TValue, typename TReturn>
+using enable_if_is_wanted = std::enable_if<is_wanted_type<TValue>, TReturn>;
+
+#define CLOE_IMPROVE_COMPILER_ERRORS 1
+
+#if defined(CLOE_IMPROVE_COMPILER_ERRORS) && (CLOE_IMPROVE_COMPILER_ERRORS > 0)

```

Comment:
Concerning philosophy B: It's not that users do not make errors, it's that these classes of errors are perhaps not occurring commonly enough to justify the guard overhead.

The fact that there is no `ifndef define endif` for the initial setting of the macro toggle indicates that we will at some point decide on one path or the other, depending on how we feel the whole experience is.

If possible, I'd like to avoid the compiler traps and guard templates. I'm more of a philosophy B person, with the caveat that I want good error messages for errors that are often made.

If we can somehow get all worlds, that would of course be ideal.

Reply:
With c++20 you get both worlds using concepts... until then, I have thought about this a little bit and perhaps something like this works:

```cpp
template <typename TValue, typename TValueFunction = typed_get_value_function_t<TValue>>
const TValueFunction* getter() const {
    static_assert(cloe::is_wanted_type<TValue>, "meaningful error message");
    validate_type<TValue>();
    const TValueFunction* get_value_fn = std::any_cast<TValueFunction>(&get_value_);
    return get_value_fn;
}
```

that way there is no enable_if, you get a "meaningful error message" upon requesting an unwanted getter and the header is probably about 150 lines shorter

Reply:
In the essence: We all agree that a static_assert inside of the body is a good solution. 

Done.